### PR TITLE
Cache AzCopy Version

### PR DESCRIPTION
**Pull request by mohsha-msft**

### CODE REVIEW

**Comment by zezha-msft** [comment]
Hmm I definitely see the advantage of caching the version info, i.e. no more waiting on the HTTP call and potential errors in locked-down environments, but how about the disadvantages?

**Comment by mohsha-msft** [comment]
> Hmm I definitely see the advantage of caching the version info, i.e. no more waiting on the HTTP call and potential errors in locked-down environments, but how about the disadvantages?

Let's look at the reason behind moving towards Cached version file.
1. Our current `latest_version.txt file` is in West US and we don't have a CDN for fast download. So for a person running AzCopy from far locations (say NewZeland or Eastern Asia), the time to complete this REST request can be upto 1.5sec. 
2. It's not a problem for non-frequent AzCopy users but for users who are running AzCopy multiple times, it is a big problem. We got an issue where a person was trying to sync 18+M blobs. Since traversal can take a lot of time, he reasonably decided to break the sync job into smaller jobs. (~4500). When he ran the commands, it took 4500 * [200ms-1s] = [16mins-80mins]. 
3. I've set the cache expiry time to be 24 hrs. So, we check for version once every day. It reduced the number of REST request to check the version once every day.

**Review by nakulkar-msft — APPROVED** [review]

**Comment by malthe** [comment]
Bump

**Comment by demoray** [comment]
Does this PR handle when azcopy is called from within a constrained environment such that the version request fails?  In some environments, this has an even bigger impact.  Ref:  https://github.com/Azure/azure-storage-azcopy/issues/1404

File: cmd/root.go [hunk 1]
```diff
@@ -212,15 +214,34 @@ func init() {
...
+		// Step 1: Fetch & validate the cached version and if it is updated, we'll return without making REST call.
+		filePath := filepath.Join(azcopyAppPathFolder, "latest_version.txt")
+		v2, err := ValidateCachedVersion(filePath)
+		if err == nil {
+			if v1.OlderThan(*v2) {
+				executablePathSegments := strings.Split(strings.Replace(os.Args[0], "\\", "/", -1), "/")
+				executableName := executablePathSegments[len(executablePathSegments)-1]
+
+				// output in info mode instead of stderr, as it was crashing CI jobs of some people
+				glcm.Info(executableName + ": A newer version " + v2.original + " is available to download\n")
```
Comment (gapra-msft) [inline cmd/root.go:239]: we should add the current version to this print string too

**Comment by gapra-msft** [comment]
Closing this PR in favor of @siminsavani-msft 's

**Pull request closed without merging**